src/state.cc

```diff
@@ -114,6 +123,17 @@ Node* State::LookupNode(StringPiece path) const {
   Paths::const_iterator i = paths_.find(path);
   if (i != paths_.end())
     return i->second;
+#ifdef _WIN32
+  string lowerpath = path.AsString();
+  transform(lowerpath.begin(), lowerpath.end(), lowerpath.begin(), ::tolower);
```

Comment:
Might it not be better to instead get the real case of the path returned? Then case sensitivity can be preserved everywhere else?

Reply:
Sorry, but I don't understand what you mean. The code first tries matching with the given casing and as a second choice also tries the lower case. The given casing can come from any input source, such as directory listing, .ninja file or .d file.

Reply:
This just blindly does lowercasing of the path. This is not really well-defined on a character-by-character basis (e.g., the classic ẞ or I → ı in Turkish cases). Instead, for each path component, ask the Windows API "what is the actual name for this component?" and build up a properly cased path. That, to me, seems like a better thing to cache than a blind lowercased pathname since it is actually likely to show up in other locations.

Plus, Windows isn't the only case insensitive platform/filesystem out there (macOS, ext4 in Linux can now be case-insensitive).

Reply:
I agree that lower case is not enough. Asking the filesystem might take time (I'm not sure) when doing for a lot of paths. What happens if a path does not exist on disk? We still want to be able to compare a generated output with an input with another casing, even if the file has not been generated yet.

Is there any other good normalizing API that can be used?

Reply:
> What happens if a path does not exist on disk?

It's assumed to be correct. Since this is mainly an issue for `/showIncludes` reporting, I think we can assume that all paths will exist at that time?

> We still want to be able to compare a generated output with an input with another casing

I think we can assume/require that generators will be sane and say that this is invalid:

```ninja
build out: mk
build uses: mk OUT
```

> Is there any other good normalizing API that can be used?

Unfortunately, I don't think so :( .

Reply:
> Since this is mainly an issue for /showIncludes reporting

Hmm. maybe that's where the "what's the real case of this path?" logic should go?

Reply:
I agree. We should assume that generators are correct and keep this to `/showIncludes`. Please feel free to create such a PR as my interest in Ninja has dropped as I am moving to Bazel.